File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/parameters/BindingSpec.java
package io.micronaut.nats.docs.parameters;

import io.micronaut.context.annotation.Property;
import io.micronaut.nats.testcontainers.Nats;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "BindingSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class BindingSpec implements TestPropertyProvider {

    @Test
    void testDynamicBinding(ProductClient productClient, ProductListener productListener) {

// tag::producer[]
        productClient.send("message body".getBytes());
        productClient.send("product", "message body2".getBytes());
// end::producer[]

        await().atMost(60, SECONDS).until(() ->
                productListener.messageLengths.size() == 2 &&
                productListener.messageLengths.contains(12) &&
                productListener.messageLengths.contains(13)
        );
    }

    @Override
    public Map<String, String> getProperties() {
        return Nats.getProperties();
    }
}

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/parameters/BindingSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="BindingSpec")
@MicronautTest(environments=["nats"])
class BindingSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_dynamic_binding(self):
        # tag::producer[]
        self.product_client.send("message body".encode())
        self.product_client.send_to_subject("product", "message body2".encode())
        # end::producer[]

        for _ in range(600):
            if len(self.product_listener.message_lengths) == 2:
                break
            sleep(0.1)
        assert sorted(self.product_listener.message_lengths) == [12, 13]
